### PR TITLE
Modify rule S2095: Expand and adjust for LaYC

### PR DESCRIPTION
**Pull request by pdschbrt**
## Review

A dedicated reviewer checked the rule description successfully for:

- [ ] logical errors and incorrect information
- [ ] information gaps and missing content
- [ ] text style and tone
- [ ] PR summary and labels follow [the guidelines](https://github.com/SonarSource/rspec/#to-modify-an-existing-rule)

### CODE REVIEW

File: rules/S2095/cfamily/rule.adoc [hunk 1]
```diff
@@ -1,43 +1,108 @@
...
+      printf("%s", buffer);
+    }
   }
-  // ...
+
   fclose(f);
-  return 0;
+  return 0; // Compliant: file `f` has been closed
 }
 ----
```
Comment (tomasz-kaminski-sonarsource) [inline rules/S2095/cfamily/rule.adoc:90]: You should extrac you example of the `File` RAII, wrapping it will match `fopen` to `fclose`.

Comment (tomasz-kaminski-sonarsource) [reply rules/S2095/cfamily/rule.adoc:90]: This also bundles operation together (construction and destruction) so make sure that `close` is matched with `fclose`. See rule about `malloc`/`delete` mismatch.

File: rules/S2095/cfamily/rule.adoc [hunk 2]
```diff
@@ -1,43 +1,108 @@
+A function call to ``++fopen++`` or ``++open++`` must be matched with a call to ``++fclose++`` or ``++close++``, respectively.
+
 == Why is this an issue?
 
-A call to the ``++fopen()++``/``++open()++`` function must be matched with a call to ``++fclose()++``/``++close++``. Otherwise, you run the risk of using up all the OS's file handles, which could lock up not just your program but potentially everything on the box.
+The standard C library provides ``++fopen++`` and the system call ``++open++`` to open and possibly create files.
+Each call to one of these functions must be matched with a respective call to ``++fclose++`` or ``++close++``.
+
+Failing to close files that have been opened may lead to using up all of the OS's file handles.
+This could lock up not just the program in question, but potentially everything that runs on the machine.
```
Comment (tomasz-kaminski-sonarsource) [inline rules/S2095/cfamily/rule.adoc:9]: I think this is just shortland for the whole next section so I would remove this sentence:
```suggestion
```

File: rules/S2095/cfamily/rule.adoc [hunk 3]
```diff
@@ -1,43 +1,108 @@
...
+This can happen if the program does not handle the error condition properly or if it enters an infinite loop trying to open files, for instance.
+In the worst case, a resource leak can lock up everything that runs on the machine.
+
+
+== How to fix it
+
+Make sure that each call to ``++fopen++`` and ``++open++`` has a matching call to ``++fclose++`` and ``++close++``, respectively.
+
+
+=== Going the extra mile
```
Comment (tomasz-kaminski-sonarsource) [inline rules/S2095/cfamily/rule.adoc:31]: This section goes after examples

File: rules/S2095/cfamily/rule.adoc [hunk 4]
```diff
@@ -1,43 +1,108 @@
...
-* https://wiki.sei.cmu.edu/confluence/x/vjdGBQ[CERT, FIO04-J.] - Release resources when they are no longer needed
-* https://wiki.sei.cmu.edu/confluence/x/QtUxBQ[CERT, FIO42-C.] - Close files when they are no longer needed
-* https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html[Try With Resources]
+=== Standards
+
+* CERT - https://wiki.sei.cmu.edu/confluence/x/vjdGBQ[FIO04-J. Release resources when they are no longer needed]
+* CERT - https://wiki.sei.cmu.edu/confluence/x/QtUxBQ[FIO42-C. Close files when they are no longer needed]
+* CWE - https://cwe.mitre.org/data/definitions/459[459 Incomplete Cleanup]
+* CWE - https://cwe.mitre.org/data/definitions/772[772 Missing Release of Resource after Effective Lifetime]
+* Java SE Tutorial - https://docs.oracle.com/javase/tutorial/essential/exceptions/tryResourceClose.html[Try With Resources]
```
Comment (tomasz-kaminski-sonarsource) [inline rules/S2095/cfamily/rule.adoc:99]: What this link does here?

**Review by tomasz-kaminski-sonarsource — CHANGES_REQUESTED** [review]

File: rules/S2095/cfamily/rule.adoc (hunk 4 above)
Comment (pdschbrt) [reply rules/S2095/cfamily/rule.adoc:99]: Will be removed, yes.

File: rules/S2095/cfamily/rule.adoc (hunk 3 above)
Comment (pdschbrt) [reply rules/S2095/cfamily/rule.adoc:30]: `Going the extra mile` is a subsection of  `How to fix it` according to the LaYC format description. This should be fine.

Comment (pdschbrt) [reply rules/S2095/cfamily/rule.adoc:30]: Sorry, I now understand your point. Will move it ;-)

File: rules/S2095/cfamily/rule.adoc [hunk 5]
```diff
@@ -34,6 +33,10 @@ Using {cpp}'s _RAII_ idiom can mitigate unmatched calls to ``++fopen++`` and ``+
 
 include::../../../shared_content/cfamily/c_file_io_raii_extra_mile.adoc[]
 
+With the design shown above, it is no longer possible to have an unmatched call to ``++fopen++``.
```
Comment (tomasz-kaminski-sonarsource) [inline rules/S2095/cfamily/rule.adoc]: It is possible if you call release. So I will say something, that this design will bey default close file that where opened, and closing it requires extra effort.

Comment (pdschbrt) [reply rules/S2095/cfamily/rule.adoc]: I've updated this now. It now matches the type of description we have for `S3588` 👍🏻

**Review by tomasz-kaminski-sonarsource — APPROVED** [review]

**Pull request merged**